File: app/src/main/java/ru/orangesoftware/orb/Query.java
package ru.orangesoftware.orb;

import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

import ru.orangesoftware.financisto.app.DependenciesHolder;
import ru.orangesoftware.financisto.utils.Logger;

public class Query<T> {
	private final Logger logger = new DependenciesHolder().getLogger();

	private final Class<T> clazz;
	private final EntityDefinition ed;
	private final SQLiteDatabase db;

	private final LinkedList<String> orderBy = new LinkedList<>();
	private String where;
	private String[] whereArgs;

	Query(EntityManager em, Class<T> clazz) {
		this.db = em.db();
		this.clazz = clazz;
		this.ed = EntityManager.getEntityDefinitionOrThrow(clazz);
	}

	public Query<T> where(Expression ex) {
		Selection s = ex.toSelection(ed);
		where = s.selection;
		whereArgs = s.selectionArgs.toArray(new String[0]);
		return this;
	}

	public Query<T> sort(Sort...sort) {
		for (Sort s : sort) {
			if (s.asc) {
				asc(s.field);
			} else {
				desc(s.field);
			}
		}
		return this;
	}

	public Query<T> asc(String field) {
		orderBy.add(ed.getColumnForField(field)+" asc");
		return this;
	}

	public Query<T> desc(String field) {
		orderBy.add(ed.getColumnForField(field)+" desc");
		return this;
	}

	public Cursor execute() {
		String query = ed.sqlQuery;
		String where = this.where;
		String[] whereArgs = this.whereArgs;
		StringBuilder sb = new StringBuilder(query);
		if (where != null) {
			sb.append(" where ").append(where);
		}
		if (!orderBy.isEmpty()) {
			sb.append(" order by ");
			boolean addComma = false;
			for (String order : orderBy) {
				if (addComma) {
					sb.append(", ");
				}
				sb.append(order);
				addComma = true;
			}
		}
		query = sb.toString();
		logger.d("QUERY %s: %s", clazz, query);
		logger.d("WHERE: %s", where != null ? where : "");
		logger.d("ARGS: %s", whereArgs != null ? Arrays.toString(whereArgs) : "");
		return db.rawQuery(query, whereArgs);
	}

	public T uniqueResult() {
		try (Cursor c = execute()) {
			if (c.moveToFirst()) {
				return EntityManager.loadFromCursor(c, clazz);
			}
			return null;
		}
	}

	public List<T> list() {
		return readEntityList(execute(), clazz);

	}

	public static <T> List<T> readEntityList(Cursor cursorRes, Class<T> clazz) {
		try (Cursor c = cursorRes) {
			List<T> list = new ArrayList<>();
			while (c.moveToNext()) {
				T e = EntityManager.loadFromCursor(c, clazz);
				list.add(e);
			}
			return list;
		}
	}
}

File: repository/src/commonMain/kotlin/ru/orangesoftware/orb/Query.kt
package ru.orangesoftware.orb

import android.database.Cursor

import androidx.sqlite.db.SupportSQLiteDatabase

import ru.orangesoftware.financisto.repository.utils.Logger

class Query<T>(
    em: EntityManager,
    private val clazz: Class<T>,
) {
    private val logger: Logger by inject()

    private val ed: ru.orangesoftware.orb.EntityDefinition = EntityManager.getEntityDefinitionOrThrow(clazz)
    private val db: SupportSQLiteDatabase = em.db()

    private val orderBy: MutableList<String> = mutableListOf()
    private var where: String? = null
    private lateinit var whereArgs: List<String>

    fun where(expression: Expression): Query<T> {
        val selection: Selection = expression.toSelection(ed)
        where = selection.selection
        whereArgs = selection.selectionArgs.toList()
        return this
    }

    fun sort(vararg sort: Sort): Query<T> {
        sort.forEach {
            if (it.asc) {
                asc(it.field)
            } else {
                desc(it.field)
            }
        }
        return this
    }

    fun asc(field: String): Query<T> {
        orderBy.add(ed.getColumnForField(field) + " asc")
        return this
    }

    fun desc(field: String): Query<T> {
        orderBy.add(ed.getColumnForField(field) + " desc")
        return this
    }

    fun execute(): Cursor {
        var query: String = ed.sqlQuery
        val where: String? = this.where
        val whereArgs: List<String>? = this.whereArgs
        val sb = StringBuilder(query)
        where?.also { sb.append(" where ").append(where) }
        if (!orderBy.isEmpty()) {
            sb.append(" order by ")
            var addComma = false
            orderBy.forEach {
                if (addComma) sb.append(", ")

                sb.append(it)
                addComma = true
            }
        }
        query = sb.toString()
        logger.d("QUERY %s: %s", clazz, query)
        logger.d("WHERE: %s", where.orEmpty())
        logger.d("ARGS: %s", whereArgs.orEmpty())
        return db.rawQuery(query, whereArgs.toTypedArray())
    }

    fun uniqueResult(): T? = execute().use {
        if (it.moveToFirst()) {
            EntityManager.loadFromCursor(it, clazz)
        } else {
            null
        }
    }

    fun list(): List<T> = readEntityList(execute(), clazz)

    companion object {
        @JvmStatic
        fun <T> readEntityList(cursorRes: Cursor, clazz: Class<T>): List<T> = cursorRes.use {
            mutableListOf<T>().apply {
                while (it.moveToNext()) {
                    EntityManager.loadFromCursor(it, clazz)?.also(::add)
                }
            }
        }
    }
}
